Add FHIR CI Support

GinoCanessa opened the pull request:
Draft PR to discuss.

GinoCanessa committed:
Add '#' style directive parsing.

GinoCanessa committed:
Spelling fixes and IDE warning cleanup.

GinoCanessa committed:
Fallback to Envrionment.GetFolderPath(UserProfile) if not found otherwise.

GinoCanessa committed:
Additions to PackageManifest to:
- capture some undocumented (but populated) properties
- add FHIR version convenience function
- add additional documentation with pointers to confluence

GinoCanessa committed:
Add extension function to fetch the highest 'latest' version from a collection of servers.

GinoCanessa committed:
WIP: Adding FhirCiClient

GinoCanessa committed:
Undo overzealous spelling corrections (revert all variable names)

GinoCanessa committed:
Pass of CI Version resolution using build timestamps as internal keys.

GinoCanessa committed:
Improvements to CI branch mangement.
Improvements to package resolution functions.

GinoCanessa committed:
Added proper date conversion for manifest.
Added installation functions to FhirCiClient.

GinoCanessa committed:
Add package deletion to IPackageCache and DiskPackageCache

GinoCanessa committed:
Add support for CI builds of FHIR Core.

mmsmits committed:
Merge branch 'develop' into fhir-ci

mmsmits commented on Firely.Fhir.Packages/Core/FhirCiClient.cs:
```diff
@@ -0,0 +1,1083 @@
+            }
+
+            // traverse the core branches to build their records
+            foreach (CiBranchRecord ciBranchRec in coreBranches)
```
This looks like a separate function to me

GinoCanessa replied:
Fair, will split those apart.

mmsmits commented on Firely.Fhir.Packages/Core/FhirCiClient.cs:
```diff
@@ -0,0 +1,1083 @@
+        /// </summary>
+        /// <param name="reference">The package reference to install or update.</param>
+        /// <param name="cache">The package cache.</param>
+        public async Task InstallOrUpdate(PackageReference reference, IPackageCache cache)
```
This actually looks like a fine function? What bothers you about it?

GinoCanessa replied:
My concern is that it does not use the existing `Install` function from `IPackageCache`.  I am not sure if that should exist there, move into an extension function, or add to the `IPackageCache` interface.

GinoCanessa committed:
Refactored CI branch to QA functions.

GinoCanessa committed:
Merge branch 'fhir-ci' of github.com:GinoCanessa/Firely.Fhir.Packages into fhir-ci

GinoCanessa committed:
Removed tools-version from manifest.  Some versions use an integer value and some use a string.

GinoCanessa committed:
FhirCiClient: made branch from repo function public (need access externally)

GinoCanessa committed:
CI client needs to delete existing content before installing - do not want removed content to merge with new content.

GinoCanessa committed:
CI Client: Fix download URL.

mmsmits committed:
Merge branch 'develop' into fhir-ci

GinoCanessa committed:
Fix: CI core builds were always pulling from the master branch.
Fix: multi-server GetLatest extension was inconsistent on .Net versions.

GinoCanessa committed:
Split convertCoreBranchesToQAs per feedback.

mmsmits committed:
Merge branch 'develop' into fhir-ci

GinoCanessa committed:
Fix: Incorrect download URLs for core FHIR CI builds.
Fix: Unit tests should download packages from build.fhir.org, even if another URL is listed as the home.
Fix: Updated build date/time for FHIR 6.0 ballot 1.

GinoCanessa committed:
Merge branch 'fhir-ci' of github.com:GinoCanessa/Firely.Fhir.Packages into fhir-ci

mmsmits commented on Firely.Fhir.Packages/Core/FhirCiClient.cs:
```diff
@@ -0,0 +1,1135 @@
+            (string? branchName, bool _) = GetBranchNameRepoLiteral(qa.RepositoryUrl);
+
+            // build the URL
+            int igUrlIndex = qa.RepositoryUrl?.IndexOf("/qa.json", StringComparison.Ordinal) ?? -1;
```
Should be a separate function

mmsmits commented on Firely.Fhir.Packages/Disk/DiskPackageCache.cs:
```diff
@@ -50,6 +50,24 @@ public async Task Install(PackageReference reference, byte[] buffer)
+        /// </summary>
+        /// <param name="reference">The package reference.</param>
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        public async Task Delete(PackageReference reference)
```
Why make it async if it isn't?

mmsmits committed:
Merge branch 'develop' into fhir-ci

mmsmits committed:
Merge branch 'develop' into fhir-ci

GinoCanessa commented:
Too far out of date - will start over when I have the chance.
